File: BOJ/11724.cpp
/*
    20:05~20:50
*/

#include <iostream>
#include <vector>
#include <cstring>

using namespace std;

int N, M, n, q, answer = 0;
int N_max = 1000;
int M_max = int((N_max*(N_max - 1))/2);

bool visited[1001];

vector<vector<int>> graph(1001);

void DFS(int node){
    int next;

    for(int i=0; i<graph[node].size(); i++){
        next = graph[node][i];
        if(!visited[next]){
            visited[next] = true;
            DFS(next);
        }
    }

    return;
}

int main(){
    cin >> N >> M;
    for(int i=0; i<M; i++){
        cin >> n >> q;

        graph[n].push_back(q);
        graph[q].push_back(n);
    }
    memset(visited, false, sizeof(bool)*(N+1));

    //visited[graph[0]] = true;
    for(int j=1; j<=N; j++){
        if(!visited[j]){
            answer += 1;
            DFS(j);
        }
    }

    cout << answer << '\n';
    return 0;
}

File: BOJ/11724.py
from collections import defaultdict
import sys
input = sys.stdin.readline

def Solve(N, edges):
    # init graph
    graph = defaultdict(list)
    for u, v in edges:
        graph[u].append(v)
        graph[v].append(u)
        
    # DFS or BFS
    visited = set()
    def DFS(node):
        stack = []
        stack.append(node)
        
        while stack:
            cnode = stack.pop()
            for nnode in graph[cnode]:
                if nnode not in visited:
                    visited.add(nnode)
                    stack.append(nnode)
        
    
    answer = 0
    for i in range(1, N+1):
        if i not in visited:
            DFS(i)
            answer += 1
        
    # output
    
    return answer


N, M = map(int, input().split())
edges = [tuple(map(int, input().split())) for _ in range(M)]
answer = Solve(N, edges)

print(answer)